source/io/JavaScriptInvoke.cpp

```diff
@@ -0,0 +1,123 @@
+*/
+
+#include "TypeDefiner.h"
+#include "ExecutionContext.h"
+#include "StringUtilities.h"
+#include "TDCodecVia.h"
+#include <stdio.h>
+
+#if defined (VIREO_TYPE_JavaScriptInvoke)
+namespace Vireo {
+
+#if kVireoOS_emscripten
+extern "C" {
+    // JavaScript function prototypes
+    // Parameters: functionName, parameters*, parametersCount, errorCheckingEnabled, errorStatus*, errorCode*, errorSource*
+    extern void jsJavaScriptInvoke(StringRef, void *, Int32, Boolean, Boolean *, Int32 *, StringRef);
+}
+#endif
+
+// Keep in synchrony with JAVASCRIPT_PARAMETER_TYPE in module_javaScriptInvoke.js
+enum JavaScriptParameterTypeEnum {
+    kJavaScriptParameterType_None = 0,
+    kJavaScriptParameterType_Int32,
+    kJavaScriptParameterType_UInt32,
+    kJavaScriptParameterType_String,
+    kJavaScriptParameterType_Boolean
+};
+
+//------------------------------------------------------------
+//! Return parameter type for the given element(index) in the parameters array
+VIREO_EXPORT Int32 Data_GetParameterType(StaticTypeAndData *parameters, Int32 index)
+{
+    TypeRef parameterType = parameters[index]._paramType;
+    Int32 type = kJavaScriptParameterType_None;
+    if (parameterType->BitEncoding() == kEncoding_UInt) {
+        type = kJavaScriptParameterType_Int32;
+    } else if (parameterType->BitEncoding() == kEncoding_S2CInt) {
+        type = kJavaScriptParameterType_UInt32;
+    } else if (parameterType->BitEncoding() == kEncoding_Boolean) {
+        type = kJavaScriptParameterType_Boolean;
+    } else if (parameterType->IsString()) {
+        type = kJavaScriptParameterType_String;
+    }
+    return type;
+}
+
+//------------------------------------------------------------
+//! Return parameter pointer for the given element(index) in the parameters array
+VIREO_EXPORT void* Data_GetParameterPointer(StaticTypeAndData *parameters, Int32 index)
+{
+    void *pData = parameters[index]._pData;
+    TypeRef parameterType = parameters[index]._paramType;
+    if (parameterType->IsString()) {
+        // We a have pointer to a StringRef, we just need the StringRef.
+        // So we can use functions that already work with StringRef on the JavaScript side.
+        pData = *(StringRef*)pData;
+    }
+    return pData;
+}
+
+//------------------------------------------------------------
+// arguments: functionName, errorCheckingEnabled, errorCluster, then variable number of inputs that can be null or any type
+struct JavaScriptInvokeParamBlock : public VarArgInstruction
+{
+    _ParamDef(StringRef, functionName);
+    _ParamDef(Boolean, errorCheckingEnabled);
+    _ParamDef(ErrorCluster, errorCluster);
+    _ParamImmediateDef(StaticTypeAndData, Parameters[1]);
+    NEXT_INSTRUCTION_METHODV()
+};
+
+extern void AddCallChainToSourceIfErrorPresent(ErrorCluster *errorCluster, ConstCStr methodName);
+extern void GenerateNotSupportedOnPlatformError(ErrorCluster *errorCluster, ConstCStr methodName);
```

Comment:
Probably should move these to a header file

Reply:
Good idea. I'll do that in a later submission.
Regarding the VIREO_REQUIRE blocks. I did that because adding a new Vireo module, changed the order of how Vireo modules are loaded and I have to add require dependencies so they load in a determined other.